### PR TITLE
Added remote log subsystem implementation.

### PR DESCRIPTION
**Pull request by wyuka**
Source commit: https://github.com/satishd/kafka/commit/54d84789c655b3268afccffa0d974213ad99cfed
Original Authors: Authors: satishd@uber.com, yingz@uber.com, kchandraprakash@uber.com

Summary:
RemoteLogManager implemented to read/copy data from/to remote tier.

- This patch contains all the required changes in core module to be able to read/fetch and write/copy log segments from remote store.
- Updated the follower fetch protocol to read only the local data available in the leader. On starting a new replica, leader epoch cache gets updated from remote store.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -860,6 +860,7 @@ project(':core') {
     implementation project(':metadata')
     implementation project(':raft')
     implementation project(':storage')
+    implementation project(':storage:api')
```
Comment (wyuka) [inline build.gradle:863]: I added this to make build succeed. This was included in the source branch from a previous commit (d8aea3fa1ee0ab399c92bb7b637309bdf6c1fc5c) which we did not merge.

File: build.gradle [hunk 2]
```diff
@@ -2224,6 +2225,7 @@ project(':jmh-benchmarks') {
     implementation project(':metadata')
     implementation project(':streams')
     implementation project(':core')
+    implementation project(':storage')
```
Comment (wyuka) [inline build.gradle:2228]: Added to make build succeed.

File: clients/src/main/java/org/apache/kafka/common/protocol/Errors.java [hunk 3]
```diff
@@ -362,7 +363,8 @@
     BROKER_ID_NOT_REGISTERED(102, "The given broker ID was not registered.", BrokerIdNotRegisteredException::new),
     INCONSISTENT_TOPIC_ID(103, "The log's topic ID did not match the topic ID in the request", InconsistentTopicIdException::new),
     INCONSISTENT_CLUSTER_ID(104, "The clusterId in the request does not match that found on the server", InconsistentClusterIdException::new),
-    TRANSACTIONAL_ID_NOT_FOUND(105, "The transactionalId could not be found", TransactionalIdNotFoundException::new);
+    TRANSACTIONAL_ID_NOT_FOUND(105, "The transactionalId could not be found", TransactionalIdNotFoundException::new),
+    OFFSET_MOVED_TO_TIERED_STORAGE(106, "The requested offset is moved to tiered storage.", OffsetMovedToTieredStorageException::new);
```
Comment (wyuka) [inline clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: Changed error number from 105 to 106, because Kafka 3.0 has an extra error `TRANSACTIONAL_ID_NOT_FOUND`

File: clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java [hunk 4]
```diff
@@ -38,6 +38,10 @@
 import org.apache.kafka.common.protocol.Errors;
 
 public class ListOffsetsRequest extends AbstractRequest {
+    /**
+     * It is used to represent the earliest message stored in the local log which is also called the local-log-start-offset
+     */
+    public static final long EARLIEST_LOCAL_TIMESTAMP = -4L;
```
Comment (wyuka) [inline clients/src/main/java/org/apache/kafka/common/requests/ListOffsetsRequest.java]: Changed integer value from `-3L` to `-4L` because Kafka 3.0 introduced `MAX_TIMESTAMP` which has value `-3L`.

File: clients/src/test/java/org/apache/kafka/common/record/RemoteLogInputStreamTest.java [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+
+    private static class RemoteLogInputStreamArgsProvider implements ArgumentsProvider {
+
+        @Override
+        public Stream<? extends Arguments> provideArguments(ExtensionContext context) {
+            List<Arguments> values = new ArrayList<>();
+            for (byte magic : asList(MAGIC_VALUE_V0, MAGIC_VALUE_V1, MAGIC_VALUE_V2)) {
+                for (CompressionType type : CompressionType.values()) {
+                    if (type == CompressionType.PASSTHROUGH)
+                        continue;
```
Comment (wyuka) [inline clients/src/test/java/org/apache/kafka/common/record/RemoteLogInputStreamTest.java:71]: Copied from `FileLogInputStreamTest`. This is done as part of the LI-HOTFIX https://github.com/linkedin/kafka/commit/e5ff962109f427d7366ffd37714501d4a2429dd1

TODO: Need to check whether Remote Storage works for MirrorMaker using passthrough compression.

File: core/src/main/scala/kafka/log/Log.scala [hunk 6]
```diff
@@ -561,6 +585,11 @@ class Log(@volatile private var _dir: File,
   /** The name of this log */
   def name  = dir.getName()
 
+  def loadProducerState(lastOffset: Long, reloadFromCleanShutdown: Boolean): Unit = lock synchronized {
```
Comment (wyuka) [inline core/src/main/scala/kafka/log/Log.scala]: This method was removed in Kafka 3.0 because of the refactoring of LogLoader. Adding this because this needs to be called from ReplicaFetcherThread.

File: core/src/main/scala/kafka/log/Log.scala [hunk 7]
```diff
@@ -561,6 +585,11 @@ class Log(@volatile private var _dir: File,
   /** The name of this log */
   def name  = dir.getName()
 
+  def loadProducerState(lastOffset: Long): Unit = lock synchronized {
```
Comment (wyuka) [inline core/src/main/scala/kafka/log/Log.scala:589]: This method was removed from Kafka 3.0 as part of the refactoring to `LogLoader`, but we added it back because it will be called from `ReplicaFetcherThread`.

File: core/src/main/scala/kafka/log/Log.scala [hunk 8]
```diff
@@ -1635,8 +1732,8 @@ class Log(@volatile private var _dir: File,
 
   private def deleteLogStartOffsetBreachedSegments(): Int = {
     def shouldDelete(segment: LogSegment, nextSegmentOpt: Option[LogSegment]): Boolean = {
-      nextSegmentOpt.exists(_.baseOffset <= logStartOffset) ||
-        (nextSegmentOpt.isEmpty && logEndOffset == logStartOffset)
+      nextSegmentOpt.exists(_.baseOffset <= localLogStartOffset) ||
+        (nextSegmentOpt.isEmpty && logEndOffset == localLogStartOffset)
```
Comment (wyuka) [inline core/src/main/scala/kafka/log/Log.scala:1742]: Added due to LI-HOTFIX https://github.com/linkedin/kafka/commit/63829efe6ba0d96ae210d0e7ae8d67bea2c650d6

File: jmh-benchmarks/src/main/java/org/apache/kafka/jmh/server/PartitionCreationBench.java [hunk 9]
```diff
@@ -138,7 +139,8 @@ public void setup() {
                 brokerTopicStats,
                 failureChannel,
                 Time.SYSTEM,
-                true);
+                true,
+                createRemoteLogManagerConfig());
```
Comment (wyuka) [inline jmh-benchmarks/src/main/java/org/apache/kafka/jmh/server/PartitionCreationBench.java:143]: Added this to make build succeed. This class was missing in 2.8. Copied changes from other benchmark diffs.

File: core/src/test/scala/unit/kafka/log/LogTestUtils.scala [hunk 10]
```diff
@@ -83,7 +83,8 @@ object LogTestUtils {
                 producerIdExpirationCheckIntervalMs: Int = LogManager.ProducerIdExpirationCheckIntervalMs,
                 lastShutdownClean: Boolean = true,
                 topicId: Option[Uuid] = None,
-                keepPartitionMetadataFile: Boolean = true): Log = {
+                keepPartitionMetadataFile: Boolean = true,
+                remoteLogEnable: Boolean = false): Log = {
```
Comment (wyuka) [inline core/src/test/scala/unit/kafka/log/LogTestUtils.scala:87]: This method has been refactored from LogTest.scala as part of Kafka 3.0. Extracted the changes that were originally made to LogTest.scala in the source commit, and applied to this file.

File: core/src/test/scala/unit/kafka/log/LogLoaderTest.scala [hunk 11]
```diff
@@ -80,7 +82,8 @@ class LogLoaderTest {
         flushCheckMs = 1000L, flushRecoveryOffsetCheckpointMs = 10000L, flushStartOffsetCheckpointMs = 10000L,
         retentionCheckMs = 1000L, maxPidExpirationMs = 60 * 60 * 1000, scheduler = time.scheduler, time = time,
         brokerTopicStats = new BrokerTopicStats, logDirFailureChannel = new LogDirFailureChannel(logDirs.size),
-        keepPartitionMetadataFile = config.usesTopicId, interBrokerProtocolVersion = config.interBrokerProtocolVersion) {
+        keepPartitionMetadataFile = config.usesTopicId, interBrokerProtocolVersion = config.interBrokerProtocolVersion,
+        remoteLogManagerConfig = remoteLogManagerConfig) {
```
Comment (wyuka) [inline core/src/test/scala/unit/kafka/log/LogLoaderTest.scala:86]: In the source commit, this change was made to LogTest.scala. In Kafka 3.0, some functionality of Log loading has been moved to `LogLoader` class. Hence, I moved the corresponding changes to `LogLoaderTest`.

File: core/src/main/scala/kafka/server/epoch/LeaderEpochFileCache.scala [hunk 12]
```diff
@@ -268,6 +282,13 @@ class LeaderEpochFileCache(topicPartition: TopicPartition,
     }
   }
 
+  def writeTo(leaderEpochCheckpoint: LeaderEpochCheckpoint): LeaderEpochFileCache = {
+    inReadLock(lock) {
+      leaderEpochCheckpoint.write(epochEntries)
+      new LeaderEpochFileCache(this.topicPartition, leaderEpochCheckpoint)
```
Comment (wyuka) [inline core/src/main/scala/kafka/server/epoch/LeaderEpochFileCache.scala:288]: The `logEndOffset` constructor argument was removed for LeaderEpochFileCache in Kafka 3.0 by this commit - https://github.com/linkedin/kafka/commit/b35704a8431a5ea99b4da3ae3b344da92c42859c

File: core/src/main/scala/kafka/server/ReplicaFetcherThread.scala [hunk 13]
```diff
@@ -384,4 +393,47 @@ class ReplicaFetcherThread(name: String,
...
+          epochs.foreach(epochEntry => {
+            log.leaderEpochCache.map(cache => cache.assign(epochEntry.epoch, epochEntry.startOffset))
+          })
+          info(s"Updated the epoch cache from remote tier till offset: $offset with size: ${epochs.size} for $partition")
+
+          // restore producer snapshot
+          val snapshotFile = Log.producerSnapshotFile(log.dir, offset)
+          Files.copy(rlm.storageManager().fetchIndex(rlsMetadata, IndexType.PRODUCER_SNAPSHOT), snapshotFile.toPath)
+          log.producerStateManager.reloadSegments()
+          log.loadProducerState(offset)
```
Comment (wyuka) [inline core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:421]: `reloadFromCleanShutdown` argument has been removed, because it is always invoked with `false`.

File: core/src/test/scala/unit/kafka/log/remote/RemoteLogManagerTest.scala [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+
+    val leaderState = new LeaderAndIsrPartitionState()
+      .setControllerEpoch(1)
+      .setLeader(brokerId)
+      .setLeaderEpoch(leaderEpoch)
+      .setIsr(Collections.singletonList(brokerId))
+      .setZkVersion(1)
+      .setReplicas(Collections.singletonList(brokerId))
+      .setIsNew(true)
+    partition.makeLeader(leaderState, new LazyOffsetCheckpoints(replicaManager.getHighWatermarkCheckpoints), None)
```
Comment (wyuka) [inline core/src/test/scala/unit/kafka/log/remote/RemoteLogManagerTest.scala:171]: In Kafka 3.0, Partition::makeLeader takes an additional argument `Option[Uuid]`, in case we want to pass topicId. This is not required for the purposes of this test, so we pass `None`.

File: core/src/main/scala/kafka/log/CleanableIndex.scala [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+abstract class CleanableIndex(@volatile var _file: File) extends Closeable {
+
+  /**
+   * Rename the file that backs this offset index
+   *
+   * @throws IOException if rename fails
+   */
+  def renameTo(toBeRenamedFile: File): Unit = {
+    try
+      if (_file.exists) Utils.atomicMoveWithFallback(_file.toPath, toBeRenamedFile.toPath, false)
```
Comment (wyuka) [inline core/src/main/scala/kafka/log/CleanableIndex.scala:33]: We are refactoring functionality from `AbstractIndex` and `TransactionIndex`, and both of them call `atomicMoveWithFallback` with the last argument `needFlushParentDir` set to `false`.

Hence, the refactored `renameTo` method should call it with `needFlushParentDir=false`. This argument was added in Kafka 3.0, please see https://github.com/linkedin/kafka/commit/66b0c5c

**Review by wyuka — COMMENTED** [review]
Left comments on each section that I have touched as part of the rebasing effort from source commit https://github.com/satishd/kafka/commit/54d84789c655b3268afccffa0d974213ad99cfed

File: clients/src/main/java/org/apache/kafka/common/protocol/Errors.java (hunk 3 above)
Comment (smccauliff) [reply clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: This has already changed in this PR.   https://github.com/apache/kafka/pull/11390  the new value is 107 for   OFFSET_MOVED_TO_TIERED_STORAGE

File: clients/src/test/java/org/apache/kafka/common/record/RemoteLogInputStreamTest.java (hunk 5 above)
Comment (smccauliff) [reply clients/src/test/java/org/apache/kafka/common/record/RemoteLogInputStreamTest.java:71]: MM/passthrough is not blocker for our internal release of tiered storage.  We can do that later.  As long as the normal produce/consume path is not broken.  We should still support the old messave versions though.  Can you add this to the list of things that might be broken for tiered storage?   You have list right?

File: clients/src/main/java/org/apache/kafka/common/utils/ChildFirstClassLoader.java [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+                            return lower.endsWith(".jar") || lower.endsWith(".zip");
+                        });
+                        if (files != null) {
+                            for (File jarFile : files) {
+                                urls.add(jarFile.getCanonicalFile().toURI().toURL());
+                            }
+                        }
+                    }
+                } catch (IOException e) {
+                }
```
Comment (smccauliff) [inline clients/src/main/java/org/apache/kafka/common/utils/ChildFirstClassLoader.java]: Can we add logging to this?  I think @sutambe has had issues with the classloader. I don't want to waste time debugging dropped exceptions.

File: core/src/main/scala/kafka/cluster/Partition.scala [hunk 17]
```diff
@@ -1158,7 +1160,7 @@ class Partition(val topicPartition: TopicPartition,
       case ListOffsetsRequest.LATEST_TIMESTAMP =>
         maybeOffsetsError.map(e => throw e)
           .orElse(Some(new TimestampAndOffset(RecordBatch.NO_TIMESTAMP, lastFetchableOffset, Optional.of(leaderEpoch))))
-      case ListOffsetsRequest.EARLIEST_TIMESTAMP =>
+      case ListOffsetsRequest.EARLIEST_TIMESTAMP | ListOffsetsRequest.EARLIEST_LOCAL_TIMESTAMP =>
         getOffsetByTimestamp
```
Comment (smccauliff) [inline core/src/main/scala/kafka/cluster/Partition.scala]: How does this search for timestamps on the remote tier?

File: clients/src/main/java/org/apache/kafka/common/protocol/Errors.java (hunk 3 above)
Comment (wyuka) [reply clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: I can change this to 107 in this PR. But we need to keep an eye on the PR, if it changes the error number by the time the PR gets accepted. In that case, we should change it here as well.

File: clients/src/main/java/org/apache/kafka/common/utils/ChildFirstClassLoader.java (hunk 16 above)
Comment (wyuka) [reply clients/src/main/java/org/apache/kafka/common/utils/ChildFirstClassLoader.java]: Good idea. I actually noticed that in the PR to merge to trunk, this exception is wrapped and thrown as a KafkaException. Doing that. Better to explicitly fail Kafka from starting if the classloader fails to load.

File: core/src/main/scala/kafka/cluster/Partition.scala (hunk 17 above)
Comment (wyuka) [reply core/src/main/scala/kafka/cluster/Partition.scala]: This calls `Log#fetchOffsetByTimestamp`. This has the following logic -

1. If ts = EARLIEST_TIMESTAMP return the logStartOffset of the log. No need to search the remote tier for this, this information is already stored inside `Log` by `RemoteLogManager` if remote tier is enabled.
2. If ts = EARLIEST_LOCAL_TIMESTAMP and remote tier is disabled, treat this as the same as the previous scenario, and return log start offset.
3. If ts = EARLIEST_LOCAL_TIMESTAMP, and remote tier is enabled, return the localLogStartOffset from the Log object.
4. If ts = a positive timestamp, then check whether the timestamp falls inside remote tier or local tier. This logic can be found inside the diff for the `Log.scala` file.

File: clients/src/test/java/org/apache/kafka/common/record/RemoteLogInputStreamTest.java (hunk 5 above)
Comment (wyuka) [reply clients/src/test/java/org/apache/kafka/common/record/RemoteLogInputStreamTest.java:71]: Yes. Noted.

File: clients/src/main/java/org/apache/kafka/common/protocol/Errors.java [hunk 18]
```diff
@@ -362,7 +363,8 @@
     BROKER_ID_NOT_REGISTERED(102, "The given broker ID was not registered.", BrokerIdNotRegisteredException::new),
     INCONSISTENT_TOPIC_ID(103, "The log's topic ID did not match the topic ID in the request", InconsistentTopicIdException::new),
     INCONSISTENT_CLUSTER_ID(104, "The clusterId in the request does not match that found on the server", InconsistentClusterIdException::new),
-    TRANSACTIONAL_ID_NOT_FOUND(105, "The transactionalId could not be found", TransactionalIdNotFoundException::new);
+    TRANSACTIONAL_ID_NOT_FOUND(105, "The transactionalId could not be found", TransactionalIdNotFoundException::new),
+    OFFSET_MOVED_TO_TIERED_STORAGE(107, "The requested offset is moved to tiered storage.", OffsetMovedToTieredStorageException::new);
```
Comment (wyuka) [inline clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: Copied this error number `107` from https://github.com/apache/kafka/pull/11390

File: clients/src/main/java/org/apache/kafka/common/utils/ChildFirstClassLoader.java [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+                            for (File jarFile : files) {
+                                urls.add(jarFile.getCanonicalFile().toURI().toURL());
+                            }
+                        }
+                    }
+                } else if (f.exists()) {
+                    urls.add(f.getCanonicalFile().toURI().toURL());
+                }
+            } catch (IOException e) {
+                throw new KafkaException(e);
```
Comment (wyuka) [inline clients/src/main/java/org/apache/kafka/common/utils/ChildFirstClassLoader.java:73]: Throwing an exception instead of swallowing this silently.

File: core/src/main/scala/kafka/server/ReplicaFetcherThread.scala [hunk 20]
```diff
@@ -230,7 +236,10 @@ class ReplicaFetcherThread(name: String,
   }
 
   override protected def fetchEarliestOffsetFromLeader(topicPartition: TopicPartition, currentLeaderEpoch: Int): Long = {
-    fetchOffsetFromLeader(topicPartition, currentLeaderEpoch, ListOffsetsRequest.EARLIEST_TIMESTAMP)
+    if (brokerConfig.interBrokerProtocolVersion >= KAFKA_3_0_IV1)
```
Comment (wyuka) [inline core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:239]: Changed from original commit, which queries with EARLIEST_LOCAL_TIMESTAMP if `IBPV >= 2.8`

**Comment by lmr3796** [comment]
nit: The commit log subject should not come with a period at the end.

**Comment by lmr3796** [comment]
nit: The commit log subject should not come with a period at the end.

File: clients/src/main/java/org/apache/kafka/common/protocol/Errors.java (hunk 18 above)
Comment (smccauliff) [reply clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: Any thoughts on how we would upgrade/downgrade a cluster which has changes to error enums, but which will potentially conflict with error enums in some future, upstream release?

Comment (wyuka) [reply clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: I agree. If a new Error enum is introduced before the PR mentioned above gets merged, and gets checked in with id `107`, we will be in trouble when upgrading.

One way out is to use a large error ID like 1001 for now, until the code gets merged.

Another alternative is that in the scenario that new error numbers emerge, we apply a hotfix to increment all subsequent error IDs so that they do not clash. That is what we do for versions for APIs `UpdateMetadata`, `LeaderAndIsr` and `StopReplicas` - we just manually increment all the version checks by 1 because we introduce an artificial version.

As a concrete example, we would change something like

```
NEW_ERROR(107, "New error description", NewError::new),
NEW_ERROR_2(108, "New error description 2", NewError2::new)
```
to
```
OFFSET_MOVED_TO_TIERED_STORAGE(107, "The requested offset is moved to tiered storage.", OffsetMovedToTieredStorageException::new),
NEW_ERROR(108, "New error description, NewError::new),
NEW_ERROR_2(109, "New error description 2", NewError2::new)
```

Comment (gitlw) [reply clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: Given the PR hasn't been merged in upstream kafka, in case 107 ends up being mapped to something else, we'd have a potential problem during the rolling upgrade. For instance, say if 107 is mapped to ERROR_X, during the rolling upgrade, the leader returning ERROR_X will be misinterpreted as OFFSET_MOVED_TO_TIERED_STORAGE by the follower.

That being said, I don't see any easy way to avoid this problem, since whatever number we use here, the same logic applies. Some hacks are possible. For instance, if we use only a temporary enum entry for now with a very high number, e.g.  LI_OFFSET_MOVED_TO_TIERED_STORAGE(1000, ....
In the future, when we pull the merged PR from upstream kafka, we can have the receiver of this error code process both error codes with the same logic, e.g.
case LI_OFFSET_MOVED_TO_TIERED_STORAGE =>
case OFFSET_MOVED_TO_TIERED_STORAGE =>
        ... process the error

after the code above is deployed, this temporary enum can be deprecated.

Comment (wyuka) [reply clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: If we add it with error code 1000 and give it a custom name, we can deprecate it's use later when the error is available in upstream.

For a while, we will have the functionality to process both errors, but emit only the upstream error. Then, with the next rolling update, we can remove it and just use the error from upstream.

Does that sound reasonable?

Comment (smccauliff) [reply clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: That sounds good.  What about other places where we have magic version numbers like the various message versions that are created?

File: core/src/main/scala/kafka/log/Log.scala [hunk 21]
```diff
@@ -1452,12 +1520,34 @@ class Log(@volatile private var _dir: File,
...
+          // We need to search the first segment whose largest timestamp is greater than the target timestamp if there is one.
+          if (earlierSegs.length < segmentsCopy.length) {
+            isFirstSegment = earlierSegs.isEmpty
+            Some(segmentsCopy(earlierSegs.length))
+          } else {
+            None
+          }
+        }
+
+        if (isFirstSegment && remoteLogManager.isDefined) {
```
Comment (gitlw) [inline core/src/main/scala/kafka/log/Log.scala:1536]: I find this condition and the logic inside this block a bit confusing.
Based on the logic below, it seems either the localOffset or the remoteOffset can be the larger value.
Under what circumstances would we have a different localOffset and remoteOffset?
Also, does this block also apply when isFirstSegment == false?

File: core/src/main/scala/kafka/log/Log.scala [hunk 22]
```diff
@@ -1896,6 +1994,11 @@ class Log(@volatile private var _dir: File,
             val deletable = logSegments.filter(segment => segment.baseOffset > targetOffset)
             removeAndDeleteSegments(deletable, asyncDelete = true, LogTruncation)
             activeSegment.truncateTo(targetOffset)
+
+            // TODO: @kamalcph check the logic again
+            this.localLogStartOffset = math.min(targetOffset, this.localLogStartOffset)
+            updateLogStartOffset(math.min(this.localLogStartOffset, this.logStartOffset))
```
Comment (gitlw) [inline core/src/main/scala/kafka/log/Log.scala]: These two lines do not align with the `truncateFullyAndStartAt` case above.
The `truncateFullyAndStartAt` calls `completeTruncatition` to update the logStartOffset or localLogStartOffset, and there is already a completeTruncation below on line 2004. So are these two lines still required?

File: clients/src/main/java/org/apache/kafka/common/protocol/Errors.java (hunk 18 above)
Comment (wyuka) [reply clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: One place I can think of is the timestamp `-4` used to signify `EARLIEST_LOCAL_TIMESTAMP` in a `ListOffsetsRequest`. We can probably change this to `-101`, and then later start treating both `-4` and `-101` the same way.

Comment (smccauliff) [reply clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: I'm ok with those solutions.  @gitlw?

Comment (smccauliff) [reply clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: Actually, do we need to make a change ApiVersion.scala?

File: core/src/main/scala/kafka/log/Log.scala (hunk 21 above)
Comment (wyuka) [reply core/src/main/scala/kafka/log/Log.scala:1536]: According to JavaDocs, the API should return the offset
> of the first message whose timestamp is greater than or equals to the target timestamp

Let us say, a ListOffsetsRequest comes in for the timestamp `1450`. We will go through the local log segments in order of their start offsets, and identify the first segment that has its highest timestamp >= 1450 *(we will refer to this as our selection criteria henceforth)*. There are some possibilities here -

## Scenario 1
### Remote Segments
```
1000-1400
1400-1410
1411-1440
```

### Segments on disk
```
1411-1440
1441-1456
1457-1480
```

In this scenario, the first log segment on disk does not fulfill our *selection criteria*, the second one does. Hence, we know for sure that there are no remote logs that can possibly meet the *selection criteria* which are not also present on the disk. We definitely know that the segment `1441-1456` contains our answer, and we rely solely on it. We don't need to check the remote tier because we know that there are no segments there that are not present on disk, and are possible candidates to have an answer.

## Scenario 2
### Remote Segments
```
1000-1400
1400-1410
1411-1440
1441-1456
```

### Segments on disk
```
1457-1480
1481-1500
```

In this scenario, the first log on disk `1457-1480` fulfills our *selection criteria*. So it is possible (but not guaranteed) for the remote tier to have some segments that have been deleted from disk that also match our *selection criteria* (in this case, `1441-1456`). In this scenario, we need to check if such a segment exists in the remote tier, and if it does, we need to check its offset.

To summarize: If the very first log segment on disk matches the *selection criteria*, then it is possible that either remote or local disk has the desired offset. If not, the local disk definitely has it. In the first scenario, it is possible to find answers both in the remote and local tier, and we should select whichever is smaller.

I too do not understand the scenario for which remoteOffset will be > localOffset when both exist. I can't think of such a scenario (maybe during unclean leader election?). Regardless, the logic sounds sound, and should just return the offset that was smaller.

We can probably check if there is scope for simplification of logic at a later point in our development. I think there is definitely some code here that can benefit from refactoring / simplification.

Comment (wyuka) [reply core/src/main/scala/kafka/log/Log.scala:1536]: We also need to test the scenario as to what happens for a ListOffsetRequests call when an unclean leader election causes epoch divergence, but RLM hasn't yet deleted the remote log segments. The call to `RLM::findOffsetByTimestamp` does not use an epoch, so is it possible that it will return an incorrect offset that has been discarded due to log divergence?

File: clients/src/main/java/org/apache/kafka/common/protocol/Errors.java (hunk 18 above)
Comment (gitlw) [reply clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: In this PR, the EARLIEST_LOCAL_TIMESTAMP enum is not exposed to the client as the client is using https://github.com/linkedin/kafka/blob/bd72ef1bf1e40feb3bc17349a385b479fa5fa530/clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java#L24 in the KafkaAdminClient.listOffsets API.
I'm not sure if this enum will be exposed to the client in a subsequent PR or in upstream kafka. If so, it'll be harder to deprecate.
If not, then I'm good with the suggested approach above.

Comment (smccauliff) [reply clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: I don't know why clients would ever use this request.  It seems like it's just issued by brokers.

File: core/src/main/scala/kafka/log/remote/RemoteLogManager.scala [hunk 23]
```diff
@@ -0,0 +1,844 @@
...
+                                                 convertToLeaderOrFollower: RLMTask => Unit): Unit = {
+    val rlmTaskWithFuture = leaderOrFollowerTasks.computeIfAbsent(topicPartition, (tp: TopicIdPartition) => {
+      val task = new RLMTask(tp)
+      // set this upfront when it is getting initialized instead of doing it after scheduling.
+      convertToLeaderOrFollower(task)
+      info(s"Created a new task: $task and getting scheduled")
+      val future = rlmScheduledThreadPool.scheduleWithFixedDelay(task, 0, delayInMs, TimeUnit.MILLISECONDS)
+      RLMTaskWithFuture(task, future)
+    })
+    convertToLeaderOrFollower(rlmTaskWithFuture.rlmTask)
```
Comment (gitlw) [inline core/src/main/scala/kafka/log/remote/RemoteLogManager.scala]: `convertToLeaderOrFollower` is already called on line 200, should it be called again here?

File: core/src/main/scala/kafka/log/remote/RemoteLogManager.scala [hunk 24]
```diff
@@ -0,0 +1,844 @@
...
+          else fetchLog(tp.topicPartition()).foreach { log =>
+            //todo-tier leader epoch to be computed here
+            val offset = remoteLogMetadataManager.highestOffsetForEpoch(tp, 0)
+            if (offset.isPresent) {
+              // todo-tier update remote index highest offset.
+              log.updateRemoteIndexHighestOffset(offset.get())
+            }
+          }
+
+          //b. cleanup/delete expired remote segments
```
Comment (gitlw) [inline core/src/main/scala/kafka/log/remote/RemoteLogManager.scala:605]: Question: is the cleanup logic still WIP?
Should the cleanup be limited to leaders of each partition?
Sometimes even after a new leader is elected, the old leader can receive the update with some delay and still thinks it's the leader, we should make sure that the code can handle the case that both the new leader and the old leader tries to delete the remote segments at the same time.

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala [hunk 25]
```diff
@@ -335,6 +356,8 @@ class BrokerTopicStats extends Logging {
       topicMetrics.closeMetric(BrokerTopicStats.FailedProduceRequestsPerSec)
       topicMetrics.closeMetric(BrokerTopicStats.TotalProduceRequestsPerSec)
       topicMetrics.closeMetric(BrokerTopicStats.ProduceMessageConversionsPerSec)
+      topicMetrics.closeMetric(BrokerTopicStats.RemoteBytesOutPerSec)
```
Comment (gitlw) [inline core/src/main/scala/kafka/server/KafkaRequestHandler.scala:405]: It seems all of the 5 new metrics are only applicable to the leader. If that's the case, should we remove all 5 of them here?

File: core/src/main/scala/kafka/log/Log.scala (hunk 22 above)
Comment (wyuka) [reply core/src/main/scala/kafka/log/Log.scala]: Discussed with Uber, we can remove the redundant check.

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala (hunk 25 above)
Comment (wyuka) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala:405]: The rest of the 3 metrics are used in followers too.

`RemoteBytesInPerSec` is emitted when bytes are copied from the remote tier to the broker. This can happen when a consumer fetches records from a broker, and the broker gets it from remote tier and responds. From Kafka 2.4.0, Fetch requests can be made to any broker part of the ISR, which can lead to `RemoteBytesInPerSec` to change. This is the same reason `BytesOutPerSec` is not in this method.

Same goes for `RemoteReadRequestsPerSec` and `FailedRemoteReadRequestsPerSec`. These can happen when a consumer fetches from any broker, and that fetches from the remote tier. This is common to leaders and followers.

File: core/src/main/scala/kafka/log/remote/RemoteLogManager.scala (hunk 24 above)
Comment (wyuka) [reply core/src/main/scala/kafka/log/remote/RemoteLogManager.scala:605]: It has been added in a subsequent commit. I think we can validate the use case you mentioned when we pull in that patch.

File: core/src/main/scala/kafka/log/remote/RemoteLogManager.scala (hunk 23 above)
Comment (wyuka) [reply core/src/main/scala/kafka/log/remote/RemoteLogManager.scala]: Will address this.

File: core/src/main/scala/kafka/log/Log.scala (hunk 21 above)
Comment (wyuka) [reply core/src/main/scala/kafka/log/Log.scala:1536]: Discussed offline. Will leave a TODO comment here with a note to validate the unclean leader election scenario during testing, or add an integration test.

File: clients/src/main/java/org/apache/kafka/common/protocol/Errors.java (hunk 18 above)
Comment (wyuka) [reply clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: We will not expose this to clients as -101. If it exposed to clients in upstream Kafka, it will be with the code -4 or -5. We can still deprecate -101 without disruption.

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala (hunk 25 above)
Comment (gitlw) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala:405]: That makes sense, thanks!

File: core/src/main/scala/kafka/log/remote/RemoteLogManager.scala (hunk 24 above)
Comment (gitlw) [reply core/src/main/scala/kafka/log/remote/RemoteLogManager.scala:605]: Thanks!

File: clients/src/main/java/org/apache/kafka/common/protocol/Errors.java (hunk 18 above)
Comment (gitlw) [reply clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: Deprecating a change in clients involves chasing who is using this, and may sometimes turn out to be a full-blown HI.
Fortunately, it's not exposed to clients now, and thus we are good.

**Review by gitlw — APPROVED** [review]
LGTM, thanks a lot for this patch!
